environment.txt
AUGUR_GITHUB_API_KEY=<PUT YOUR GITHUB API KEY HERE>
AUGUR_GITHUB_USERNAME=Mbaoma 
AUGUR_GITLAB_API_KEY=<PUT YOUR GITLAB KEY HERE>
AUGUR_GITLAB_USERNAME=<PUT YOUR GITLAB USERNAME HERE>
AUGUR_DB=<PUT YOUR DATABASE CONNECTION STRING HERE>
AUGUR_RABBITMQ_USERNAME=<OPTIONAL - SET DESIRED RABBITMQ USER>
AUGUR_RABBITMQ_PASSWORD=<OPTIONAL - SET DESIRED RABBITMQ PASSWORD>


docs/source/deployment/nginx-configuration.rst
Web Server Configuration
------------------------

Configuring nginx for Augur to run behind nginx requires you to have certain options available for symlinks and other basic nginx options. The `nginx.conf` file below is one example of a configuration known to work.

Once you have nginx configured, run these commands to make sure everything is loaded and configured correctly:

1. ``sudo nginx -t`` to make sure it's configured correctly.
	`nginx: the configuration file /etc/nginx/nginx.conf syntax is ok`
	`nginx: configuration file /etc/nginx/nginx.conf test is successful`
2. ``sudo systemctl restart nginx`` on Ubuntu.
3. sudo nginx on OS X.

------------------
Server Compilation
------------------

**Your Augur instance must compile with a publicly accessible domain that the frontend instance will be able to access.**

1. Your ``augur.config.json`` server block **must** be set like this:

.. code-block:: json

	{
	    "Server": {
	        "cache_expire": 3600,
	        "host": "subdomain.domain.tld",
	        "port": "5000",
	        "workers": 8
	    }
    }

2.   Compile Augur (this wires the host and port into the frontend so people pulling the web pages of Augur, in the `frontend/` subdirectory are referring to the right endpoints for this instance.): ``make rebuild``
3.   Run Augur: ``nohup augur backend start >augur.log 2>augur.err &``


------------------
nginx
------------------

------------------
nginx.conf
------------------

.. code-block::

	user www-data;
	worker_processes auto;
	pid /run/nginx.pid;
	include /etc/nginx/modules-enabled/*.conf;

	worker_rlimit_nofile 30000;

	events {
		worker_connections 768;
		# multi_accept on;
	}

	http {

		##
		# Basic Settings
		##
		disable_symlinks off;

		sendfile on;
		tcp_nopush on;
		tcp_nodelay on;
		keepalive_timeout 65;
		types_hash_max_size 2048;
		# server_tokens off;

		 server_names_hash_bucket_size 64;
		# server_name_in_redirect off;

		include /etc/nginx/mime.types;
		default_type application/octet-stream;

		##
		# SSL Settings
		##

		ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3; # Dropping SSLv3, ref: POODLE
		ssl_prefer_server_ciphers on;

		##
		# Logging Settings
		##

		access_log /var/log/nginx/access.log;
		error_log /var/log/nginx/error.log;

		##
		# Gzip Settings
		##

		gzip on;

		# gzip_vary on;
		# gzip_proxied any;
		# gzip_comp_level 6;
		# gzip_buffers 16 8k;
		# gzip_http_version 1.1;
		# gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript;

		##
		# Virtual Host Configs
		##

		include /etc/nginx/conf.d/*.conf;
		include /etc/nginx/sites-enabled/*;
	}

--------------------
Site Configuration
--------------------

This file will be located in the ``/etc/nginx/sites-enabled`` directory on most Linux distributions.  Mac OSX keeps these files in the ``/usr/local/etc/nginx/sites-enabled`` directory. **Note that Augur's backend server must be running**

.. code-block::

		server {
		        listen 80;
		        server_name  <<your server subdomain.domain.tld>>;

		        root /home/user/.../<<augur-instance-home>>/frontend/dist;
		        index index.html index.htm;

		        location / {
		        root /home/user/.../<<augur-instance-home>>/frontend/dist;
		        try_files $uri $uri/ /index.html;
		        }

		#        location /api/unstable/ {
		#                proxy_pass http://census.osshealth.io:5000;
		#                proxy_set_header Host $host;
		#        }

		        location /api_docs/ {
		        root /home/user/.../<<augur-instance-home>>/frontend/dist;
		        index index.html;
		        }


		        error_log /var/log/nginx/augur.censusscienceosshealth.error.log;
		        access_log /var/log/nginx/augur.censusscienceosshealth.access.log;

		}
    
--------------------
Enabling HTTPS
--------------------

HTTPS is an extension of HTTP. It is used for secure communications over a computer networks by encrypting your data so it is not vulnerable to MIM(Man-in-the-Middle) attacks etc. While Augur's API data might not be very sensitive, it would still be a nice feature to have so something can't interfere and provide wrong data. Additionally, the user may not feel very comfortable using an application when the browser is telling the user it is not secure. Features such as logins is an example of information that would be particularly vulnerable to attacks. Lastly, search engine optimization actually favors applications on HTTPS over HTTP.

This guide will start on a fully configured EC2 Ubuntu 20.04 instance, meaning it is assumed to already have Augur installed and running with all of its dependencies(PostgreSQL, Nginx, etc).

~~~~~~~~~~~~~~~~~~~~~~~~~~~
Let's Encrypt/Certbot
~~~~~~~~~~~~~~~~~~~~~~~~~~~

The easiest way to get an HTTPS server up is to make use of `Let's Encrypt <https://letsencrypt.org/>`_'s `Certbot <https://certbot.eff.org/>`_ tool. It is an open source tool that is so good and it will even alter the nginx configuration for you automatically to enable HTTPS. Following their guide for ``Ubuntu 20.04``, run ``sudo snap install --classic certbot``, ``sudo ln -s /snap/bin/certbot /usr/bin/certbot``, and then ``sudo certbot --nginx``.

.. code-block:: console

	# Example Certificate Response Using Certbot

	Which names would you like to activate HTTPS for?
	- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	1: augur.augurlabs.io
	2: new.augurlabs.io
	3: old.augurlabs.io
	4: augur.chaoss.io
	- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	Select the appropriate numbers separated by commas and/or spaces, or leave input
	blank to select all options shown (Enter 'c' to cancel): 4
	Requesting a certificate for augur.chaoss.io

	Successfully received certificate.
	Certificate is saved at: /etc/letsencrypt/live/augur.chaoss.io/fullchain.pem
	Key is saved at:         /etc/letsencrypt/live/augur.chaoss.io/privkey.pem
	This certificate expires on 2022-07-12.
	These files will be updated when the certificate renews.
	Certbot has set up a scheduled task to automatically renew this certificate in the background.

	Deploying certificate
	Successfully deployed certificate for augur.chaoss.io to /etc/nginx/sites-enabled/augur.chaoss.io
	Congratulations! You have successfully enabled HTTPS on https://augur.chaoss.io

	- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	If you like Certbot, please consider supporting our work by:
	 * Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
	 * Donating to EFF:                    https://eff.org/donate-le


~~~~~~~~~~~~~~~~~~~
Fixing the Backend
~~~~~~~~~~~~~~~~~~~

Now our server is configured properly and our frontend is being served over HTTPS, but there's an extra problem: the backend APIs are still being served over HTTP resulting in a ``blocked loading mixed active content`` error. This issue is a deep rooted issue and several files need to be modified to accomodate HTTPS.

First, we will start with lines 29, 33, & 207 of ``augur/frontend/src/AugurAPI.ts`` and rewrite the URL to use the HTTPS protocol instead of HTTP. We will then do this again in ``augur/frontend/src/common/index.tx`` & ``augur/frontend/src/compare/index.ts`` where the ``AugurAPI`` constructor was called and passed an HTTP protocol. Next we need to configure gunicorn in the backend to support our SSL certificates, but by default certbot places these in a directory that requires root access. Copy these files by running ``sudo cp /etc/letsencrypt/live/<server name here>/fullchain.pem /home/ubuntu/augur/fullchain.pem`` and ``sudo cp /etc/letsencrypt/live/<server name here>/privkey.pem /home/ubuntu/augur/privkey.pem`` into augur's root directory, then change the user and group permissions with ``sudo chown ubuntu <filename.pem>`` and ``sudo chgrp ubuntu <filename.pem`` for both pem files. Now that the user permissions are set properly, gunicorn should be able to access them but we still need to add them to our gunicorn configuration document in ``augur/application.py``. Change the corresponding code block to look like this:

.. code-block:: python

    self.gunicorn_options = {
            'bind': '%s:%s' % (self.config.get_value("Server", "host"), self.config.get_value("Server", "port")),
            'workers': int(self.config.get_value('Server', 'workers')),
            'timeout': int(self.config.get_value('Server', 'timeout')),
            'certfile': '/home/ubuntu/augur/fullchain.pem',
            'keyfile': '/home/ubuntu/augur/privkey.pem'
        }




augur/application/config.py
import sqlalchemy as s
from sqlalchemy import and_, update
import json
from typing import List, Any, Optional
import os
from augur.application.db.models import Config 
from augur.application.db.util import execute_session_query

def get_development_flag_from_config(engine):
    
    from logging import getLogger
    from augur.application.db.session import DatabaseSession

    logger = getLogger(__name__)
    with DatabaseSession(logger, engine=engine) as session:

        config = AugurConfig(logger, session)

        section = "Augur"
        setting = "developer"

        flag = config.get_value(section, setting)

    return flag

def get_development_flag(engine):

    return os.getenv("AUGUR_DEV") or get_development_flag_from_config(engine) or False



default_config = {
            "Augur": {
                "developer": 0,
                "version": 1
            },
            "Keys": {
                "github": "<gh_api_key>",
                "gitlab": "<gl_api_key>"
            },
            #TODO: a lot of these are deprecated.
            "Facade": {
                "check_updates": 1,
                "create_xlsx_summary_files": 1,
                "delete_marked_repos": 0,
                "fix_affiliations": 1,
                "force_invalidate_caches": 1,
                "limited_run": 0,
                "multithreaded": 1,
                "nuke_stored_affiliations": 0,
                "pull_repos": 1,
                "rebuild_caches": 1,
                "run_analysis": 1,
                "run_facade_contributors": 1
            },
            "Server": {
                "cache_expire": "3600",
                "host": "0.0.0.0",
                "port": 5000,
                "workers": 6,
                "timeout": 6000,
                "ssl": False,
                "ssl_cert_file": None, 
                "ssl_key_file": None 
            },
            "Logging": {
                "logs_directory": "",
                "log_level": "INFO",
            },
            "Celery": {
                "worker_process_vmem_cap": 0.25,
                "refresh_materialized_views_interval_in_days": 1
            },
            "Redis": {
                "cache_group": 0, 
                "connection_string": "redis://127.0.0.1:6379/"
            },
            "RabbitMQ": {
                "connection_string": "amqp://augur:password123@localhost:5672/augur_vhost"
            },
            "Tasks": {
                "collection_interval": 30
            },
            "Message_Insights": {
                    "insight_days": 30,
                    "models_dir": "message_models"
            },
            "Clustering_Task": {
                "max_df": 0.9,
                "max_features": 1000,
                "min_df": 0.1,
                "num_clusters": 4
            },
            "Insight_Task": {
                # TODO: How to store metrics in database config?
                "confidence_interval": 95,
                "contamination": 0.1,
                "switch": 1,
                "workers": 1,
                "training_days": 1000,
                "anomaly_days": 14
            },
            "Task_Routine": {
                "prelim_phase": 1,
                "primary_repo_collect_phase": 1,
                "secondary_repo_collect_phase": 1,
                "facade_phase": 1,
                "machine_learning_phase": 0
            }
        }


def convert_type_of_value(config_dict, logger=None):
        
        data_type = config_dict["type"]

        if data_type == "str" or data_type is None:
            return config_dict

        elif data_type == "int":
            config_dict["value"] = int(config_dict["value"])

        elif data_type == "bool":
            value = config_dict["value"]
            
            if value.lower() == "false":
                config_dict["value"] = False
            else:
                config_dict["value"] = True

        elif data_type == "float":
            config_dict["value"] = float(config_dict["value"])

        else:
            if logger:
                logger.error(f"Need to add support for {data_type} types to config") 
            else:
                print(f"Need to add support for {data_type} types to config")

        return config_dict

class AugurConfig():

    from augur.application.db.session import DatabaseSession

    session: DatabaseSession

    def __init__(self, logger, session: DatabaseSession):

        self.session = session
        self.logger = logger

        self.accepted_types = ["str", "bool", "int", "float", "NoneType"]
        self.default_config = default_config

    def get_section(self, section_name: str) -> dict:
        """Get a section of data from the config.

        Args:
            section_name: The name of the section being retrieved

        Returns:
            The section data as a dict
        """
        query = self.session.query(Config).filter_by(section_name=section_name)
        section_data = execute_session_query(query, 'all')
        
        section_dict = {}
        for setting in section_data:
            setting_dict = setting.__dict__

            setting_dict = convert_type_of_value(setting_dict, self.logger)

            setting_name = setting_dict["setting_name"]
            setting_value = setting_dict["value"]

            section_dict[setting_name] = setting_value

        return section_dict


    def get_value(self, section_name: str, setting_name: str) -> Optional[Any]:
        """Get the value of a setting from the config.

        Args:
            section_name: The name of the section that the setting belongs to 
            setting_name: The name of the setting

        Returns:
            The value from config if found, and None otherwise
        """

        # TODO temporary until added to the DB schema
        if section_name == "frontend" and setting_name == "pagination_offset":
            return 25

        try:
            query = self.session.query(Config).filter(Config.section_name == section_name, Config.setting_name == setting_name)
            config_setting = execute_session_query(query, 'one')
        except s.orm.exc.NoResultFound:
            return None

        setting_dict = config_setting.__dict__

        setting_dict = convert_type_of_value(setting_dict, self.logger)

        return setting_dict["value"]


    def load_config(self) -> dict:
        """Get full config as a dictionary.
        
        Returns:
            The config from the database
        """
        # get all the sections in the config table
        query = self.session.query(Config.section_name)
        section_names = execute_session_query(query, 'all')

        config = {}
        # loop through and get the data for each section
        for section_name in section_names:

            section_data = self.get_section(section_name[0])

            # rows with a section of None are on the top level, 
            # so we are adding these values to the top level rather 
            # than creating a section for them
            if section_name[0] is None:
                for key in list(section_data.keys()):
                    config[key] = section_data[key]
                continue

            # add section data to config object
            config[section_name[0]] = section_data

        return config


    def empty(self) -> bool:
        """Determine if a config is empty.
        
        Returns:
            True if the config is empty, and False if it is not
        """
        query = self.session.query(Config)
        return execute_session_query(query, 'first') is None

    def is_section_in_config(self, section_name: str) -> bool:
        """Determine if a section is in the config.
        
        Args:
            section_name: section to search for in config

        Returns:
            True if section is in the config, and False if it is not
        """
        query = self.session.query(Config).filter(Config.section_name == section_name)
        return execute_session_query(query, 'first') is not None


    def add_or_update_settings(self, settings: List[dict]):
        """Add or update a list of settings.

        Args:
            list of settings with dicts containing section_name, setting_name, value, and optionally type

        Examples:
            type is optional
            setting = {
                    "section_name": section_name,
                    "setting_name": setting_name,
                    "value": value,
                    "type": data_type # optional
                }
        """
        for setting in settings:

            if "type" not in setting:
                setting["type"] = setting["value"].__class__.__name__

            if setting["type"] == "NoneType":
                setting["type"] = None

        #print(f"\nsetting: {settings}")
        #self.session.insert_data(settings,Config, ["section_name", "setting_name"])

            #Check if setting exists.
            query = self.session.query(Config).filter(and_(Config.section_name == setting["section_name"],Config.setting_name == setting["setting_name"]) )

            if execute_session_query(query, 'first') is None:
                self.session.insert_data(setting,Config, ["section_name", "setting_name"])
            else:
                #If setting exists. use raw update to not increase autoincrement
                update_query = (
                    update(Config)
                    .where(Config.section_name == setting["section_name"])
                    .where(Config.setting_name == setting["setting_name"])
                    .values(value=setting["value"])
                )

                self.session.execute(update_query)
                self.session.commit()
       

    def add_section_from_json(self, section_name: str, json_data: dict) -> None:
        """Add a section from a dict.
        
        Args:
            section_name: The name of the section being added
            json_data: The data being added
        """
        data_keys = list(json_data.keys())

        settings = []
        for key in data_keys:

            value = json_data[key]

            if isinstance(value, dict) is True:
                # TODO: Uncomment out when insights worker config stuff is resolved
                # self.logger.error(f"Values cannot be of type dict: {value}")
                return

            setting = {
                "section_name": section_name,
                "setting_name": key,
                "value": json_data[key],
            }
            settings.append(setting)

        self.add_or_update_settings(settings)


    def load_config_file(self, file_path: str) -> dict:
        """Add a section from a dict.
        
        Args:
            file_path: Path to json file being loaded

        Returns:
            data in the json file
        """
        with open(file_path, 'r') as f:
            file_data = json.load(f)

            return file_data

    def load_config_from_dict(self, dict_data: dict) -> None:
        """Create config from a dict.
        
        Args:
            dict_data: The data being loaded into the config
        """
        section_names = list(dict_data.keys())

        for section_name in section_names:
            
            value = dict_data[section_name]
            #print(f"\n{value}")
            # check for "sections" that are actually just a key value pair 
            # and not a key that has a value of type dict
            if isinstance(value, dict) is True:
                self.add_section_from_json(section_name=section_name, json_data=value)

            else:
                self.logger.error(f"Error! {section_name}: {value} will not be added because a section must have a dict as its values (all of the top level keys in the config must have a value of type dict")

    def clear(self) -> None:
        """Remove all values from the config."""
        self.session.query(Config).delete()
        self.session.commit()

    def remove_section(self, section_name: str) -> None:
        """Remove a section from the config.
        
        Args:
            section_name: The name of the section being deleted
        """
        self.session.query(Config).filter(Config.section_name == section_name).delete()
        self.session.commit()


    def create_default_config(self) -> None:
        """Create default config in the database."""
        self.load_config_from_dict(self.default_config)
